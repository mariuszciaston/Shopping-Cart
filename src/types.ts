export type ProductTypes = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductsContextType = {
  productsData: ProductTypes[] | null;
  error: null;
  loading: boolean;
};
