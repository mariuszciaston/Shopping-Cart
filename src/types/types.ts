export type Product = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductState = {
  productsData: Product[] | null;
  error: string | null;
  loading: boolean;
};

export type QuantitySelectorProps = {
  quantity: number;
  setQuantity: (quantity: number) => void;
};

export type CartItem = {
  id: number;
  quantity: number;
};

export type CartState = {
  itemsInCart: CartItem[];
  setItemsInCart: React.Dispatch<React.SetStateAction<CartItem[]>>;
};

export type UpdateCart = (
  id: CartItem["id"],
  quantity: CartItem["quantity"],
  itemsInCart: CartState["itemsInCart"],
  setItemsInCart: CartState["setItemsInCart"],
) => void;

export type ShoppingCart = {
  productsData: Product[] | null;
  itemsInCart: CartItem[];
};
