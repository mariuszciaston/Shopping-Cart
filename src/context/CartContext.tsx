import { CartContextType } from "@/types/types";
import { createContext, useContext } from "react";

export const CartContext = createContext<CartContextType | null>(null);

export const useCartContext = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }
  return context;
};
