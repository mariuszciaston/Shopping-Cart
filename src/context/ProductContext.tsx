import { ProductsContextType } from "@/types";
import { createContext, useContext } from "react";

export const ProductsContext = createContext<ProductsContextType | null>(null);

export const useProductsContext = () => {
  const context = useContext(ProductsContext);
  if (!context) {
    throw new Error(
      "useProductsContext must be used within a ProductsProvider",
    );
  }
  return context;
};
