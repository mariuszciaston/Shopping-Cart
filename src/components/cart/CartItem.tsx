import { Link } from "react-router-dom";
import { Trash2 } from "lucide-react";
import { useCartContext } from "@/context/CartContext";
import QuantitySelector from "@/components/ui/QuantitySelector";
import { Product } from "@/types/types";

function CartItem({ product }: { product: Product }) {
  const { itemsInCart, setItemsInCart } = useCartContext();

  return (
    <div className="flex flex-1 flex-wrap items-center justify-between gap-4">
      <div className="flex items-center gap-4 sm:min-w-[480px] sm:flex-1">
        <Link to={`/products/${product.id}`}>
          <img
            src={product.image}
            alt={product.title}
            className="aspect-square w-32 min-w-32 rounded-md border bg-white object-contain p-4"
          />
        </Link>

        <Link to={`/products/${product.id}`}>
          <p className="flex flex-col">
            <span className="font-bold">{product.title}</span>${product.price}
          </p>
        </Link>
      </div>

      <div className="flex flex-1 items-center gap-4">
        <QuantitySelector
          quantity={
            itemsInCart.find((item) => item.id === product.id)?.quantity ?? 0
          }
          setQuantity={(newQuantity) => {
            const updatedCart = itemsInCart.map((item) =>
              item.id === product.id
                ? { ...item, quantity: newQuantity }
                : item,
            );
            setItemsInCart(updatedCart);
          }}
        />

        <Trash2
          className="cursor-pointer hover:text-red-500"
          onClick={() => {
            const updatedCart = itemsInCart.filter(
              (item) => item.id !== product.id,
            );
            setItemsInCart(updatedCart);
          }}
        />

        <p className="flex flex-1 justify-end">
          $
          {(
            (itemsInCart.find((item) => item.id === product.id)?.quantity ??
              0) * product.price
          ).toFixed(2)}
        </p>
      </div>
    </div>
  );
}

export default CartItem;
