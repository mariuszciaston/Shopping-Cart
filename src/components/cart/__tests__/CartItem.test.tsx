import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { beforeEach, describe, expect, it, vi, Mock } from "vitest";
import { useCartContext } from "@/context/CartContext";
import CartItem from "../CartItem";
import { Product } from "@/types/types";

vi.mock("@/context/CartContext", () => ({
  useCartContext: vi.fn(),
}));

const mockUseCartContext = useCartContext as Mock;

const product: Product = {
  id: 1,
  title: "Sample Product",
  price: 19.99,
  image: "sample-image.jpg",
  description: "description",
};

const itemsInCart = [{ id: 1, quantity: 3 }];

describe("CartItem component", () => {
  beforeEach(() => {
    mockUseCartContext.mockReturnValue({
      itemsInCart,
      setItemsInCart: vi.fn(),
    });
  });

  it("renders product information correctly", () => {
    render(
      <BrowserRouter>
        <CartItem product={product} />
      </BrowserRouter>,
    );

    expect(screen.getByRole("img")).toHaveAttribute("src", "sample-image.jpg");
    expect(screen.getByTestId("product-title")).toHaveTextContent(
      "Sample Product",
    );
    expect(screen.getByTestId("product-price")).toHaveTextContent("$19.99");
  });

  it("shows quantity correctly", () => {
    render(
      <BrowserRouter>
        <CartItem product={product} />
      </BrowserRouter>,
    );

    expect(screen.getByTestId("quantity-input")).toHaveValue(3);
  });

  it("updates quantity correctly by value edit", () => {
    const setItemsInCart = vi.fn();
    mockUseCartContext.mockReturnValue({
      itemsInCart,
      setItemsInCart,
    });

    render(
      <BrowserRouter>
        <CartItem product={product} />
      </BrowserRouter>,
    );

    const quantityInput = screen.getByTestId("quantity-input");

    fireEvent.change(quantityInput, { target: { value: 4 } });
    expect(setItemsInCart).toHaveBeenCalledWith([{ id: 1, quantity: 4 }]);

    fireEvent.change(quantityInput, { target: { value: 5 } });
    expect(setItemsInCart).toHaveBeenCalledWith([{ id: 1, quantity: 5 }]);
  });

  it("removes item from cart when Trash icon is clicked", () => {
    const setItemsInCart = vi.fn();
    mockUseCartContext.mockReturnValue({
      itemsInCart,
      setItemsInCart,
    });

    render(
      <BrowserRouter>
        <CartItem product={product} />
      </BrowserRouter>,
    );

    const trashIcon = screen.getByTestId("remove-item");
    fireEvent.click(trashIcon);

    expect(setItemsInCart).toHaveBeenCalledWith([]);
  });

  it("displays total price correctly", () => {
    render(
      <BrowserRouter>
        <CartItem product={product} />
      </BrowserRouter>,
    );

    expect(screen.getByTestId("sum")).toHaveTextContent("$59.97");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <CartItem product={product} />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
