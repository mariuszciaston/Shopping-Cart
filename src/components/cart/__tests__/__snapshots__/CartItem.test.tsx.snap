// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartItem component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-1 flex-wrap items-center justify-between gap-4"
  >
    <div
      class="flex items-center gap-4 sm:min-w-[480px] sm:flex-1"
    >
      <a
        data-discover="true"
        href="/products/1"
      >
        <img
          alt="Sample Product"
          class="aspect-square w-32 min-w-32 rounded-md border bg-white object-contain p-4"
          src="sample-image.jpg"
        />
      </a>
      <a
        data-discover="true"
        href="/products/1"
      >
        <p
          class="flex flex-col"
        >
          <span
            class="font-bold"
            data-testid="product-title"
          >
            Sample Product
          </span>
          <span
            data-testid="product-price"
          >
            $19.99
          </span>
        </p>
      </a>
    </div>
    <div
      class="flex flex-1 items-center gap-4"
    >
      <div
        class="flex gap-2"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
          data-testid="quantity-lower"
        >
          -
        </button>
        <input
          class="flex h-9 rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm m-none w-[4rem] pr-0 text-center"
          data-testid="quantity-input"
          max="99"
          min="1"
          type="number"
          value="3"
        />
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
          data-testid="quantity-boost"
        >
          +
        </button>
      </div>
      <svg
        class="lucide lucide-trash2 cursor-pointer hover:text-red-500"
        data-testid="remove-item"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 6h18"
        />
        <path
          d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
        />
        <path
          d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
      <p
        class="flex flex-1 justify-end"
        data-testid="sum"
      >
        $59.97
      </p>
    </div>
  </div>
</DocumentFragment>
`;
